sys/include/xtimer.h

```diff
@@ -389,6 +393,75 @@ int xtimer_msg_receive_timeout64(msg_t *msg, uint64_t us);
 #define XTIMER_USLEEP_UNTIL_OVERHEAD 10
 #endif
 
+#ifndef XTIMER_HZ
+/**
+ * @brief Tick rate of the underlying hardware timer.
+ *
+ * This value is used to convert between hardware ticks and microseconds.
+ *
+ * This is supposed to be defined per-device in e.g., periph_conf.h.
+ */
+#define XTIMER_HZ (1000000ul)
+#endif
+
+/* Some optimizations for common timer frequencies */
+#if XTIMER_HZ == (1000000ul)
+/* This is the most straightforward as the xtimer API is based around
+ * microseconds for representing time values. */
+#define _xtimer_us_to_ticks(us)    (us) /* no-op */
+#define _xtimer_ticks_to_us(ticks) (ticks) /* no-op */
+#define _xtimer_us_to_ticks64(us)    (us) /* no-op */
+#define _xtimer_ticks_to_us64(ticks) (ticks) /* no-op */
+
+#elif XTIMER_HZ == (32768ul)
+/* This is a common frequency for RTC crystals. We use the fact that the
+ * greatest common divisor between 32768 and 1000000 is 64, so instead of
+ * multiplying by the fraction (32768 / 1000000), we will instead use
+ * (512 / 15625), which reduces the truncation caused by the integer widths */
+
+inline static uint32_t _xtimer_us_to_ticks(uint32_t us) {
+    /* return (usec * 512) / 15625; */
+    /*
+     * Optimization to avoid division instructions or division libraries.
+     * This code gives equivalent machine code on ARMv7-M (Cortex-M3, M4) as
+     * generated by compiling the above code line.
+     * at -O2 or higher with GCC 4.9. Cortex-M4 with optimization level -Os or
+     * lower does not automatically perform this optimization.
+     * Cortex-M0 (ARMv6-M) does not have hardware division and will insert a
+     * call to the helper function __aeabi_uidiv which is quite complex and
+     * slow, and uses an additional 200 bytes of ROM compared to the helper
+     * function required for multiplying two 32 bit integers into a 64 bit result.
+     */
+    return ((uint64_t)(us << 9) * 0x431bde83ul) >> (12 + 32);
```

Comment:
Executing the conversion as sparsely as possible was the main reason I chose to put it inside the public functions and make xtimer use ticks internally. 

Also, I wanted to keep the API design around microseconds to avoid conversion macros inside the calls similar to what the hwtimer API used to require before. In my opinion it is good to have a standard unit defined for the timer delays, especially when writing portable code between platforms. 

Both the macro approach and the static inline function approach lets the compiler optimize away unnecessary conversions by precomputing the result when using constants for offsets (useful in device drivers which require delays which are specified in the data sheet of the device).